docs/specs/buffer.md

```diff
@@ -0,0 +1,31 @@
+# Buffer Specification
+
+This document specifies Vector's buffer behavior for the development of Vector.
+
+The key words “MUST”, “MUST NOT”, “REQUIRED”, “SHALL”, “SHALL NOT”, “SHOULD”,
+“SHOULD NOT”, “RECOMMENDED”, “MAY”, and “OPTIONAL” in this document are to be
+interpreted as described in [RFC 2119].
+
+<!-- MarkdownTOC autolink="true" style="ordered" indent="   " -->
+
+1. [Instrumentation](#instrumentation)
+
+<!-- /MarkdownTOC -->
+
+## Instrumentation
```

Comment:
Should we add the same [intro paragraph](https://github.com/vectordotdev/vector/blob/master/docs/specs/component.md#instrumentation) from the component spec here? Are you planning to use events to drive instrumentation for the buffer?

Reply:
I discussed with @jszwedko, and we decided to omit the `internal events` framework used by components to avoid dependencies between the `vector-core` implementation of buffers and `vector`. [Following this precedent in the code](https://github.com/vectordotdev/vector/blob/30ff3da8c3ee14428096db33931aa89c14c59326/lib/vector-core/buffers/src/acker.rs#L30-L38).

Reply:
I see, thanks. What is the ideal solution here? That comment seems like a quick workaround. It seems like [`InternalEvent`](https://github.com/vectordotdev/vector/blob/b5105d9fb0a2a7d29c562b183456dc31cfaf5825/src/internal_events/mod.rs#L285) could also be moved to `vector-core`?

cc @blt

Reply:
I'd be curious to know what the actual plan is, straight use of `metrics` crate? 

I believe we could easily move the trait down into core. `InternalEvents` strikes me as a relatively clean concept, in that implementations are allowed to vary how they actual function; so I don't see any need to adjust the trait prior to move.

Reply:
Yep, originally planned to use the macros from `metrics` directly. Moving `InternalEvents` seems like a more robust solution that helps avoid discrepancies in internal instrumentation. Will update the spec.

Reply:
👍 to moving the trait down and macro down into `vector-core`.